README.md

# ArduinoCI Ruby gem (`arduino_ci`)
[![Gem Version](https://badge.fury.io/rb/arduino_ci.svg)](https://rubygems.org/gems/arduino_ci)
[![Documentation](http://img.shields.io/badge/docs-rdoc.info-blue.svg)](http://www.rubydoc.info/gems/arduino_ci/1.3.0)
[![Gitter](https://badges.gitter.im/Arduino-CI/arduino_ci.svg)](https://gitter.im/Arduino-CI/arduino_ci?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge)
[![GitHub Marketplace](https://img.shields.io/badge/Get_it-on_Marketplace-informational.svg)](https://github.com/marketplace/actions/arduino_ci)

Arduino CI tests [Arduino libraries](https://arduino.github.io/arduino-cli/library-specification/); it was created to enable better collaboration among Arduino library maintainers and contributors, by enabling automated code checks to be performed as part of a pull request process.

* enables running unit tests against the library **without hardware present**
* provides a system of mocks that allow fine-grained control over the hardware inputs, including the system's clock
* verifies compilation of any example sketches included in the library
* can test a wide range of arduino boards with different hardware options available
* can be run both locally and as part of CI (GitHub Actions, TravisCI, Appveyor, etc.)
* runs on multiple platforms -- any platform that supports the Arduino IDE
* provides detailed analysis of segfaults in compilers that support such debugging features

> Note: for running tests in response to [GitHub events](https://docs.github.com/en/free-pro-team@latest/developers/webhooks-and-events/github-event-types), an [Arduino CI GitHub Action](https://github.com/marketplace/actions/arduino_ci) is available for your convenience.  This method of running `arduino_ci` is driven by Docker, which may also serve your local testing needs (as it does not require a ruby environment to be installed).

Arduino CI works on multiple platforms, which should enable your CI system of choice to leverage it for testing.

Platform | CI Status
---------|:---------
OSX      | [![OSX Build Status](https://github.com/Arduino-CI/arduino_ci/workflows/macos/badge.svg)](https://github.com/Arduino-CI/arduino_ci/actions?workflow=macos)
Linux    | [![Linux Build Status](https://github.com/Arduino-CI/arduino_ci/workflows/linux/badge.svg)](https://github.com/Arduino-CI/arduino_ci/actions?workflow=linux)
Windows  | [![Windows Build status](https://github.com/Arduino-CI/arduino_ci/workflows/windows/badge.svg)](https://github.com/Arduino-CI/arduino_ci/actions?workflow=windows)


## Quick Start

This project has the following dependencies:

* `ruby` 2.5 or higher
* A compiler like `g++` (on OSX, `clang` works; on Cygwin, use the `mingw-gcc-c++` package)
* `python` (if using a board architecture that requires it, e.g. ESP32, ESP8266; see [this issue](https://github.com/Arduino-CI/arduino_ci/issues/235#issuecomment-739629243)). Consider `pyserial` as well.

In that environment, you can install by running `gem install arduino_ci`.  To update to a latest version, use `gem update arduino_ci`.

You can now test your library by simply running the command `arduino_ci.rb` from your library directory.  This will perform the following:

* validation of some fields in `library.properties`, if it exists
* running unit tests from files found in `test/`, if they exist
* testing compilation of example sketches found in `examples/`, if they exist

### Assumptions About Your Repository

Arduino expects all libraries to be in a specific `Arduino/libraries` directory on your system.  If your library is elsewhere, `arduino_ci` will _automatically_ create a symbolic link in the `libraries` directory that points to the directory of the project being tested.  This simplifieds working with project dependencies, but **it can have unintended consequences on Windows systems**.

> If you use a Windows system **it is recommended that you only run `arduino_ci` from project directories that are already inside the `libraries` directory** because [in some cases deleting a folder that contains a symbolic link to another folder can cause the _entire linked folder_ to be removed instead of just the link itself](https://superuser.com/a/306618).

### Changes to Your Repository

Unit testing binaries created by `arduino_ci` should not be committed to the codebase.  To avoid that, add the following to your `.gitignore`:

```ignore-list
# arduino_ci unit test binaries and artifacts
*.bin
*.bin.dSYM
```

### A Quick Example

For a fairly minimal practical example of a unit-testable library repo that you can copy from, see [the `Arduino-CI/Blink` repository](https://github.com/Arduino-CI/Blink).


## Advanced Start

New features and bugfixes reach GitHub before they reach a released ruby gem.  Alternately, it may be that (for your own reasons) you do not wish to install `arduino_ci` globally on your system.  A few additional setup steps are required if you wish to do this.

### You Need Ruby _and_ Bundler

In addition to version 2.5 or higher, you'll also need to `gem install bundler` to a minimum of version 2.0 if it's not already there.  You may find it easiest to do this by using [`rbenv`](https://github.com/rbenv/rbenv).

You will need to add a file called `Gemfile` (no extension) to your Arduino project.

#### Non-root installation

If you are simply trying to avoid the need to install `arduino_ci` system-wide (which may require administrator permissions), your `Gemfile` would look like this:

```ruby
source 'https://rubygems.org'

# Replace 1.2 with the desired version of arduino_ci.  See https://guides.rubygems.org/patterns/#pessimistic-version-constraint
gem 'arduino_ci', '~> 1.2'
```

It would also make sense to add the following to your `.gitignore`:
```ignore-list
/.bundle/
vendor
```

> Note: this used to be the recommended installation method, but with the library's maturation it's better to avoid the use of `Gemfile` and `bundle install` by just installing as per the "Quick Start" instructions above.


#### Using the latest-available code

If you want to use the latest code on GitHub, your `Gemfile` would look like this:

```ruby
source 'https://rubygems.org'

# to use the latest github code in a given repo and branch, replace the below values for git: and ref: as needed
gem 'arduino_ci', git: 'https://github.com/ArduinoCI/arduino_ci.git', ref: '<your desired ref, branch, or tag>'
```


#### Using a version of `arduino_ci` source code on your local machine

First, Thanks!  See [CONTRIBUTING.md](CONTRIBUTING.md).  Your `Gemfile` would look like this:

```ruby
source 'https://rubygems.org'

gem 'arduino_ci', path: '/path/to/development/dir/for/arduino_ci'
```


### Installing the Dependencies

Fulfilling the `arduino_ci` library dependencies is as easy as running these two
commands:

```console
$ bundle config set --local path 'vendor/bundle'
$ bundle install
```

This will create a `Gemfile.lock` in your project directory, which you may optionally check into source control.  A broader introduction to ruby dependencies is outside the scope of this document.


### Running `arduino_ci.rb` To Test Your Library

With that installed, just the following shell command each time you want the tests to execute:

```console
$ bundle exec arduino_ci.rb
```


### Reference

For more information on the usage of `arduino_ci.rb`, see [REFERENCE.md](REFERENCE.md).  It contains information such as:

* How to configure build options (platforms to test, Arduino library dependencies to install) with an `.arduino-ci.yml` file
* Where to put unit test files
* How to structure unit test files
* How to control the global (physical) state of the Arduino board
* How to modify the Arduino platforms, compilers, test plans, etc


## Setting up Pull Request Testing and/or External CI

> **Note:** `arduino_ci.rb` expects to be run from the root directory of your Arduino project library.

### Arduino CI's Own GitHub action

[![GitHub Marketplace](https://img.shields.io/badge/Get_it-on_Marketplace-informational.svg)](https://github.com/marketplace/actions/arduino_ci)


### Your Own Scripted GitHub Action

GitHub Actions allows you to automate your workflows directly in GitHub.
No additional steps are needed.
Just create a YAML file with the information below in your repo under the `.github/workflows/` directory.

```yaml
on: [push, pull_request]
jobs:
  runTest:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: ruby/setup-ruby@v1
        with:
          ruby-version: 2.6
      - run: |

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [Unreleased]
### Added

### Changed

### Deprecated

### Removed

### Fixed
- `arduino_ci.rb --help` no longer crashes

### Security


## [1.4.0] - 2022-12-28
### Added
- Allow use of watchdog timer in application code (though it doesn't do anything)
- Show output from successful compile
- `--min-free-space=N` command-line argument to fail if free space is below required value
- Add `_BV()` macro.
- Support for `dtostrf()`
- Added a CI workflow to lint the code base
- Added a CI workflow to check for spelling errors
- Extraction of byes usage in a compiled sketch is now calculated in a method: `ArduinoBackend.last_bytes_usage`
- Added ```nano_every``` platform to represent ```arduino:megaavr``` architecture
- Working directory is now printed in test runner output
- Explicitly include `irb` via rubygems

### Changed
- We now compile a shared library to be used for each test.
- Put build artifacts in a separate directory to reduce clutter.
- Replace `#define yield() _NOP()` with `inline void yield() { _NOP(); }` so that other code can define a `yield()` function.
- Update .gitattributes so we have consistent line endings
- Change 266 files from CRLF to LF.
- Run tests on push as well as on a pull request so developers can see impact
- `ArduinoBackend` now exposes `config_file_path` instead of `config_dir` so that we can be explicit about [strange behavior in `arduino-cli` that isn't going to change anytime soon](https://github.com/arduino/arduino-cli/issues/753)
- Use `arduino-cli` version `0.29.0` as the backend
- Test runner detects console width if possible, allowing variable width from 80-132 chars

### Fixed
- Properly report compile errors in GitHub Actions.
- Fix copy/paste error to allow additional warnings for a platform
- Apply "rule of three" to Client copy constructor and copy assignment operator
- Run Windows tests on Windows not Ubuntu
- Properly report error in building shared library
- A missing `examples` directory no longer causes a crash in `cpp_library.rb`
- Referring to an undefined platform no longer causes a crash; it's now a helpful error message
- A copy/paste error that prevented compiler warning flags from being supplied has been fixed, via jgfoster
- RSpec was not communicating compile errors from unit test executables that failed to build. Now it does, via jgfoster
- Windows paths now avoid picking up backslashes, for proper equality comparisons
- Libraries are now considered installed if their entry is a symlink (for which `exist?` would return `false`)


## [1.3.0] - 2021-01-13
### Added
- Better indications of the build phases in the test runner `arduino_ci.rb`
- Better indications of which example sketch is being compiled as part of testing

### Changed
- Topmost installation instructions now suggest `gem install arduino_ci` instead of using a `Gemfile`.  Reasons for using a `Gemfile` are listed and discussed separately further down the README.
- Stream::readStreamUntil() no longer returns delimiter

### Removed
- scanning of `library.properties`; this can and should now be performed by the standalone [`arduino-lint` tool](https://arduino.github.io/arduino-lint).

### Fixed
- Example sketches with no configured platforms were printing the wrong configuration values to the debug message
- Libraries directory was not being automatically created prior to attempting to change directory into it
- A style error whose "fix" caused an _actual_ error.
- Proper installation order for NetworkLib in CI workflow configuration


## [1.2.0] - 2021-01-06
### Added
- Environment variable to run a custom initialization script during CI testing: `CUSTOM_INIT_SCRIPT`
- Environment variable to run from a subdirectory during CI testing: `USE_SUBDIR`
- `assertComparativeEquivalent()` and `assertComparativeNotEquivalent()` to evaluate equality on an `a - b == 0` basis (and/or `!(a > b) && !(a < b)`)
- `assertEqualFloat()` and `assertNotEqualFloat()` for comparing floats with epsilon
- `assertInfinity()` and `assertNotInfinity()` for comparing floats to infinity

exe/arduino_ci.rb
#!/usr/bin/env ruby
require 'arduino_ci'
require 'set'
require 'pathname'
require 'optparse'
require 'io/console'

# be flexible between 80 and 132 cols of output
WIDTH = begin
  [132, [80, IO::console.winsize[1] - 2].max].min
rescue NoMethodError
  80
end
VAR_CUSTOM_INIT_SCRIPT = "CUSTOM_INIT_SCRIPT".freeze
VAR_USE_SUBDIR         = "USE_SUBDIR".freeze
VAR_EXPECT_EXAMPLES    = "EXPECT_EXAMPLES".freeze
VAR_EXPECT_UNITTESTS   = "EXPECT_UNITTESTS".freeze
VAR_ARDUINO_CI_PRE_UNIT_TEST_RUN_SCRIPT  = "ARDUINO_CI_PRE_UNIT_TEST_RUN_SCRIPT".freeze
VAR_ARDUINO_CI_POST_UNIT_TEST_RUN_SCRIPT = "ARDUINO_CI_POST_UNIT_TEST_RUN_SCRIPT".freeze

@failure_count = 0
@passfail = proc { |result| result ? "✓" : "✗" }
@backend = nil

# Use some basic parsing to allow command-line overrides of config
class Parser
  def self.parse(options)
    unit_config = {}
    output_options = {
      skip_unittests: false,
      skip_compilation: false,
      ci_config: {
        "unittest" => unit_config
      },
      min_free_space: nil,
    }

    opt_parser = OptionParser.new do |opts|
      opts.banner = "Usage: #{File.basename(__FILE__)} [options]"

      opts.on("--skip-unittests", "Don't run unit tests") do |p|
        output_options[:skip_unittests] = p
      end

      opts.on("--skip-examples-compilation", "Don't compile example sketches") do |p|
        output_options[:skip_compilation] = p
      end

      opts.on("--testfile-select=GLOB", "Unit test file (or glob) to select") do |p|
        unit_config["testfiles"] ||= {}
        unit_config["testfiles"]["select"] ||= []
        unit_config["testfiles"]["select"] << p
      end

      opts.on("--testfile-reject=GLOB", "Unit test file (or glob) to reject") do |p|
        unit_config["testfiles"] ||= {}
        unit_config["testfiles"]["reject"] ||= []
        unit_config["testfiles"]["reject"] << p
      end

      opts.on("--min-free-space=VALUE", "Minimum free SRAM memory for stack/heap, in bytes") do |p|
        output_options[:min_free_space] = p.to_i
      end

      opts.on("-h", "--help", "Prints this help") do
        puts opts
        puts
        puts "Additionally, the following environment variables control the script:"
        puts " - #{VAR_CUSTOM_INIT_SCRIPT} - if set, this script will be run from the Arduino/libraries directory"
        puts "       prior to any automated library installation or testing (e.g. to install unofficial libraries)"
        puts " - #{VAR_CUSTOM_INIT_SCRIPT_SHELL} - if set, this will override the"
        puts "       default shell (/bin/sh) used to execute #{VAR_CUSTOM_INIT_SCRIPT} with."
        puts " - #{VAR_ARDUINO_CI_PRE_UNIT_TEST_RUN_SCRIPT} and/or #{VAR_ARDUINO_CI_POST_UNIT_TEST_RUN_SCRIPT}"
        puts "       if set, run the script before/after each unit test run"
        puts " - #{VAR_USE_SUBDIR} - if set, the script will install the library from this subdirectory of the cwd"
        puts " - #{VAR_EXPECT_EXAMPLES} - if set, testing will fail if no example sketches are present"
        puts " - #{VAR_EXPECT_UNITTESTS} - if set, testing will fail if no unit tests are present"
        exit
      end
    end

    opt_parser.parse!(options)
    output_options
  end
end

# Read in command line options and make them read-only
@cli_options = (Parser.parse ARGV).freeze

def print_backend_logs
  puts "========== Last backend command (if relevant):"
  puts @backend.last_msg.to_s
  puts "========== Backend Stdout:"
  puts @backend.last_out
  puts "========== Backend Stderr:"
  puts @backend.last_err
end

# terminate after printing any debug info.  TODO: capture debug info
def terminate(final = nil)
  puts "Failures: #{@failure_count}"
  print_backend_logs unless @failure_count.zero? || final || @backend.nil?
  retcode = @failure_count.zero? ? 0 : 1
  exit(retcode)
end

# make a nice status line for an action and react to the action
# TODO / note to self: inform_multiline is tougher to write
#   without altering the signature because it only leaves space
#   for the checkmark _after_ the multiline, it doesn't know how
#   to make that conditionally the body
# @param message String the text of the progress indicator
# @param multiline boolean whether multiline output is expected
# @param mark_fn block (string) -> string that says how to describe the result
# @param on_fail_msg String custom message for failure
# @param tally_on_fail boolean whether to increment @failure_count
# @param abort_on_fail boolean whether to abort immediately on failure (i.e. if this is a fatal error)
def perform_action(message, multiline, mark_fn, on_fail_msg, tally_on_fail, abort_on_fail)
  line = "#{message}... "
  endline = "...#{message} "
  if multiline
    puts line
  else
    print line
  end
  $stdout.flush
  result = yield
  mark = mark_fn.nil? ? "" : mark_fn.call(result)
  # if multiline, put checkmark at full width
  print endline if multiline
  puts mark.to_s.rjust(WIDTH - line.length, " ")
  unless result
    puts on_fail_msg unless on_fail_msg.nil?
    @failure_count += 1 if tally_on_fail
    # print out error messaging here if we've captured it
    terminate if abort_on_fail
  end
  result
end

# Make a nice status for something that defers any failure code until script exit
def attempt(message, &block)
  perform_action(message, false, @passfail, nil, true, false, &block)
end

# Make a nice status for something that defers any failure code until script exit
def attempt_multiline(message, &block)
  perform_action(message, true, @passfail, nil, true, false, &block)
end

# Make a nice status for something that kills the script immediately on failure
FAILED_ASSURANCE_MESSAGE = "This may indicate a problem with your configuration; halting here".freeze
def assure(message, &block)
  perform_action(message, false, @passfail, FAILED_ASSURANCE_MESSAGE, true, true, &block)
end

def assure_multiline(message, &block)
  perform_action(message, true, @passfail, FAILED_ASSURANCE_MESSAGE, true, true, &block)
end

def inform(message, &block)
  perform_action(message, false, proc { |x| x }, nil, false, false, &block)
end

def inform_multiline(message, &block)
  perform_action(message, true, nil, nil, false, false, &block)
end

def rule(char)
  puts char[0] * WIDTH
end

def warn(message)
  inform("WARNING") { message }
end

def phase(name)
  puts
  rule("=")
  inform("Beginning the next phase of testing") { name }
end

def banner
  art = [
    "         .                  __  ___",
    " _, ,_  _| , . * ._   _    /  `  | ",
    "(_| [ `(_] (_| | [ ) (_)   \\__. _|_   v#{ArduinoCI::VERSION}",
  ]

  pad = " " * ((WIDTH - art[2].length) / 2)
  art.each { |l| puts "#{pad}#{l}" }
  puts
end

# Assure that a platform exists and return its definition
def assured_platform(purpose, name, config)
  platform_definition = config.platform_definition(name)
  assure("Requested #{purpose} platform '#{name}' is defined in 'platforms' YML") { !platform_definition.nil? }
  platform_definition
end

# Return true if the file (or one of the dirs containing it) is hidden
def file_is_hidden_somewhere?(path)
  # this is clunkly but pre-2.2-ish ruby doesn't return ascend as an enumerator
  path.ascend do |part|
    return true if part.basename.to_s.start_with? "."
  end
  false
end

# print out some files
def display_files(pathname)
  # `find` doesn't follow symlinks, so we should instead
  realpath = ArduinoCI::Host.symlink?(pathname) ? ArduinoCI::Host.readlink(pathname) : pathname

  # suppress directories and dotfile-based things
  all_files = realpath.find.select(&:file?)
  non_hidden = all_files.reject { |path| file_is_hidden_somewhere?(path) }

  # print files with an indent
  puts "  Files (excluding hidden files): #{non_hidden.size}"
  non_hidden.each { |p| puts "    #{p}" }
end

# @return [Array<String>] The list of installed libraries
def install_arduino_library_dependencies(library_names, on_behalf_of, already_installed = [])
  installed = already_installed.clone
  (library_names.map { |n| @backend.library_of_name(n) } - installed).each do |l|
    if l.installed?
      inform("Using pre-existing dependency of #{on_behalf_of}") { l.name }
    else
      assure("Installing dependency of #{on_behalf_of}: '#{l.name}'") do
        next nil unless l.install

        l.name
      end
    end
    installed << l.name
    installed += install_arduino_library_dependencies(l.arduino_library_dependencies, l.name, installed)
  end
  installed
end

# @param platforms [Array<String>] list of platforms to consider
# @param specific_config [CIConfig] configuration to use
def install_all_packages(platforms, specific_config)

  # get packages from platforms
  all_packages = specific_config.platform_info.select { |p, _| platforms.include?(p) }.values.map { |v| v[:package] }.compact.uniq

  all_packages.each do |pkg|
    next if @backend.boards_installed?(pkg)

    url = assure("Board package #{pkg} has a defined URL") { specific_config.package_url(pkg) }
    @backend.board_manager_urls = [url]
    assure("Installing board package #{pkg}") { @backend.install_boards(pkg) }
  end
end

# @param expectation_envvar [String] the name of the env var to check
# @param operation [String] a description of what operation we might be skipping
# @param filegroup_name [String] a description of the set of files without which we effectively skip the operation
# @param dir_description [String] a description of the directory where we looked for the files
# @param dir [Pathname] the directory where we looked for the files
def handle_expectation_of_files(expectation_envvar, operation, filegroup_name, dir_description, dir_path)
  # alert future me about running the script from the wrong directory, instead of doing the huge file dump
  # otherwise, assume that the user might be running the script on a library with no actual unit tests
  if Pathname.new(__dir__).parent == Pathname.new(Dir.pwd)
    inform_multiline("arduino_ci seems to be trying to test itself") do
      [
        "arduino_ci (the ruby gem) isn't an arduino project itself, so running the CI test script against",
        "the core library isn't really a valid thing to do... but it's easy for a developer (including the",
        "owner) to mistakenly do just that.  Hello future me, you probably meant to run this against one of",
        "the sample projects in SampleProjects/ ... if not, please submit a bug report; what a wild case!"
      ].each { |l| puts "  #{l}" }
      false
    end
    exit(1)
  end

  # either the directory is empty, or it doesn't exist at all. message accordingly.
  (problem, dir_desc, dir) = if dir_path.exist?
    ["No #{filegroup_name} were found in", dir_description, dir_path]
  else
    ["No #{dir_description} at", "base directory", dir_path.parent]
  end

  inform(problem) { dir_path }
  explain_and_exercise_envvar(expectation_envvar, operation, "contents of #{dir_desc}") { display_files(dir) }
end

# @param expectation_envvar [String] the name of the env var to check
# @param operation [String] a description of what operation we might be skipping
# @param block_desc [String] a description of what information will be dumped to assist the user
# @param block [Proc] a function that dumps information
def explain_and_exercise_envvar(expectation_envvar, operation, block_desc, &block)
  inform("Environment variable #{expectation_envvar} is") { "(#{ENV[expectation_envvar].class}) #{ENV[expectation_envvar]}" }
  if ENV[expectation_envvar].nil?
    inform_multiline("Skipping #{operation}") do
      puts "  In case that's an error, displaying #{block_desc}:"
      block.call
      puts "  To force an error in this case, set the environment variable #{expectation_envvar}"
      true
    end
  else
    assure_multiline("Displaying #{block_desc} before exit") do
      block.call
      false
    end
  end
end

# report and return the set of compilers
def get_annotated_compilers(config, cpp_library)
  # check GCC
  compilers = config.compilers_to_use
  assure("The set of compilers (#{compilers.length}) isn't empty") { !compilers.empty? }
  compilers.each do |gcc_binary|
    attempt_multiline("Checking #{gcc_binary} version") do
      version = cpp_library.gcc_version(gcc_binary)
      next nil unless version

      puts version.split("\n").map { |l| "    #{l}" }.join("\n")
      version
    end
    inform("libasan availability for #{gcc_binary}") { cpp_library.libasan?(gcc_binary) }
  end
  compilers
end

# Run custom custom script specified by user.
#
# This feature is to drive GitHub actions / docker image installation where the container is
# in a clean-slate state but needs some way to have custom library versions injected into it.
# In this case, the user provided script would fetch a git repo or some other method.
def run_custom_script(env_var, *args)
  script_path = ENV[env_var]
  script_shell = ENV[env_var + "_SHELL"] || "/bin/sh"
  inform("Environment variable #{env_var}") { "'#{script_path}'" }
  return if script_path.nil?
  return if script_path.empty?

  script_pathname = Pathname.getwd + script_path
  assure("Script at #{env_var} exists") { script_pathname.exist? }

  assure_multiline("Running #{script_pathname} with #{script_shell} in libraries working dir") do
    Dir.chdir(@backend.lib_dir) do
      IO.popen([script_shell, script_pathname.to_s, *args], err: [:child, :out]) do |io|
        io.each_line { |line| puts "    #{line}" }
      end
    end
  end
end

# Auto-select some platforms to test based on the information available
#
# Top choice is always library.properties -- otherwise use the default.
# But filter that through any non-default config
#
# @param config [CIConfig] the overridden config object
# @param reason [String] description of why we might use this platform (i.e. unittest or compilation)
# @param desired_platforms [Array<String>] the platform names specified
# @param library_properties [Hash] the library properties defined by the library
# @return [Array<String>] platforms to use
def choose_platform_set(config, reason, desired_platforms, library_properties)

  # if there are no properties or no architectures, defer entirely to desired platforms
  if library_properties.nil? || library_properties.architectures.nil? || library_properties.architectures.empty?
    # verify that all platforms exist
    desired_platforms.each { |p| assured_platform(reason, p, config) }
    return inform_multiline("No architectures listed in library.properties, using configured platforms") do
      desired_platforms.each { |p| puts "    #{p}" } # this returns desired_platforms
    end
  end

  if library_properties.architectures.include?("*")
    return inform_multiline("Wildcard architecture in library.properties, using configured platforms") do
      desired_platforms.each { |p| puts "    #{p}" } # this returns desired_platforms
    end
  end

  platform_architecture = config.platform_info.transform_values { |v| v[:board].split(":")[1] }
  supported_platforms = platform_architecture.select { |_, a| library_properties.architectures.include?(a) }

  if config.is_default
    # completely ignore default config, opting for brute-force library matches
    # OTOH, we don't need to assure platforms because we defined them
    return inform_multiline("Default config, platforms matching architectures in library.properties") do
      supported_platforms.keys.each do |p|  # rubocop:disable Style/HashEachMethods
        puts "    #{p}"
      end # this returns supported_platforms
    end
  end

  desired_supported_platforms = supported_platforms.select { |p, _| desired_platforms.include?(p) }.keys
  desired_supported_platforms.each { |p| assured_platform(reason, p, config) }
  inform_multiline("Configured platforms that match architectures in library.properties") do
    desired_supported_platforms.each do |p|
      puts "    #{p}"
    end # this returns supported_platforms
  end
end

# Unit test procedure
def perform_unit_tests(cpp_library, file_config)
  phase("Unit testing")
  if @cli_options[:skip_unittests]
    inform("Skipping unit tests") { "as requested via command line" }
    return
  end

  config = file_config.with_override_config(@cli_options[:ci_config])
  compilers = get_annotated_compilers(config, cpp_library)

  inform("Library conforms to Arduino library specification") { cpp_library.one_point_five? ? "1.5" : "1.0" }

  # Handle lack of test files
  if cpp_library.test_files.empty?
    handle_expectation_of_files(VAR_EXPECT_UNITTESTS, "unit tests", "test files", "tests directory", cpp_library.tests_dir)
    return
  end

  # Get platforms, handle lack of them
  platforms = choose_platform_set(config, "unittest", config.platforms_to_unittest, cpp_library.library_properties)
  if platforms.empty?
    explain_and_exercise_envvar(VAR_EXPECT_UNITTESTS, "unit tests", "platforms and architectures") do
      puts "    Configured platforms: #{config.platforms_to_unittest}"
      puts "    Configuration is default: #{config.is_default}"
      arches = cpp_library.library_properties.nil? ? nil : cpp_library.library_properties.architectures
      puts "    Architectures in library.properties: #{arches}"
    end
  end

  # having undefined platforms is a config error
  platforms.select { |p| config.platform_info[p].nil? }.each do |p|
    assure("Platform '#{p}' is defined in configuration files") { false }
  end

  install_arduino_library_dependencies(config.aux_libraries_for_unittest, "<unittest/libraries>")

  platforms.each do |p|
    puts
    compilers.each do |gcc_binary|
      run_custom_script(VAR_ARDUINO_CI_PRE_UNIT_TEST_RUN_SCRIPT, p, gcc_binary)
      # before compiling the tests, build a shared library of everything except the test code
      build_result = build_shared_library(gcc_binary, p, config, cpp_library)
      @failure_count += 1 unless build_result
      next unless build_result

      # now build and run each test using the shared library build above
      config.allowable_unittest_files(cpp_library.test_files).each do |unittest_path|
        unittest_name = unittest_path.basename.to_s
        puts "--------------------------------------------------------------------------------"
        attempt_multiline("Unit testing #{unittest_name} with #{gcc_binary} for #{p}") do
          exe = cpp_library.build_for_test(unittest_path, gcc_binary)
          puts
          unless exe
            puts "Last command: #{cpp_library.last_cmd}"
            puts cpp_library.last_out
            puts cpp_library.last_err
            next false
          end
          cpp_library.run_test_file(exe)
        end
      end
      run_custom_script(VAR_ARDUINO_CI_POST_UNIT_TEST_RUN_SCRIPT, p, gcc_binary)
    end
  end
end

def build_shared_library(gcc_binary, platform, config, cpp_library)
  attempt_multiline("Build shared library with #{gcc_binary} for #{platform}") do
    exe = cpp_library.build_shared_library(
      config.aux_libraries_for_unittest,
      gcc_binary,
      config.gcc_config(platform)
    )
    puts
    unless exe
      puts "Last command: #{cpp_library.last_cmd}"
      puts cpp_library.last_out
      puts cpp_library.last_err
    end
    exe
  end
end

def perform_example_compilation_tests(cpp_library, config)
  phase("Compilation of example sketches")
  if @cli_options[:skip_compilation]
    inform("Skipping compilation of examples") { "as requested via command line" }
    return
  end

  library_examples = cpp_library.example_sketches

  if library_examples.empty?
    handle_expectation_of_files(VAR_EXPECT_EXAMPLES, "builds", "examples", "the examples directory", cpp_library.examples_dir)
    return
  end

  library_examples.each do |example_path|
    example_name = File.basename(example_path)
    puts
    inform("Discovered example sketch") { example_name }

    ovr_config = config.from_example(example_path)
    platforms = choose_platform_set(ovr_config, "library example", ovr_config.platforms_to_build, cpp_library.library_properties)

    # having no platforms defined is probably an error
    if platforms.empty?
      explain_and_exercise_envvar(VAR_EXPECT_EXAMPLES, "examples compilation", "platforms and architectures") do
        puts "    Configured platforms: #{ovr_config.platforms_to_build}"
        puts "    Configuration is default: #{ovr_config.is_default}"
        arches = cpp_library.library_properties.nil? ? nil : cpp_library.library_properties.architectures
        puts "    Architectures in library.properties: #{arches}"
      end
    end

    # having undefined platforms is a config error
    platforms.select { |p| ovr_config.platform_info[p].nil? }.each do |p|
      assure("Platform '#{p}' is defined in configuration files") { false }
    end

    install_all_packages(platforms, ovr_config)
    install_arduino_library_dependencies(ovr_config.aux_libraries_for_build, "<compile/libraries>")

    platforms.each do |p|
      board = ovr_config.platform_info[p][:board] # assured to exist, above
      attempt("Compiling #{example_name} for #{board}") do
        ret = @backend.compile_sketch(example_path, board)
        unless ret
          puts "Last command: #{@backend.last_msg}"
          puts @backend.last_err
        end
        ret
      end

      next if @cli_options[:min_free_space].nil?

      usage = @backend.last_bytes_usage
      min_free_space = @cli_options[:min_free_space]
      attempt("Checking that free space of #{usage[:free]} is less than desired minimum #{min_free_space}") do
        min_free_space <= usage[:free]
      end
    end
  end
end

banner
inform("Host OS") { ArduinoCI::Host.os }
inform("Working directory") { Dir.pwd }

# initialize command and config
config = ArduinoCI::CIConfig.default.from_project_library
@backend = ArduinoCI::ArduinoInstallation.autolocate!
inform("Located arduino-cli binary") { @backend.binary_path.to_s }
if @backend.lib_dir.exist?
  inform("Found libraries directory") { @backend.lib_dir }
else
  assure("Creating libraries directory") { @backend.lib_dir.mkpath || true }
end

# run any library init scripts from the library itself.
run_custom_script(VAR_CUSTOM_INIT_SCRIPT)

# initialize library under test
inform("Environment variable #{VAR_USE_SUBDIR}") { "'#{ENV[VAR_USE_SUBDIR]}'" }
cpp_library_path = Pathname.new(ENV[VAR_USE_SUBDIR].nil? ? "." : ENV[VAR_USE_SUBDIR])
cpp_library = assure("Installing library under test") do
  @backend.install_local_library(cpp_library_path)
end

# Warn if the library name isn't obvious
assumed_name = @backend.name_of_library(cpp_library_path)
ondisk_name = cpp_library_path.realpath.basename.to_s
warn("Installed library named '#{assumed_name}' has directory name '#{ondisk_name}'") if assumed_name != ondisk_name

if !cpp_library.nil?
  inform("Library installed at") { cpp_library.path.to_s }
else
  # this is a longwinded way of failing, we aren't really "assuring" anything at this point
  assure_multiline("Library installed successfully") do
    puts @backend.last_msg
    false
  end
end

install_arduino_library_dependencies(
  cpp_library.arduino_library_dependencies,
  "<#{ArduinoCI::CppLibrary::LIBRARY_PROPERTIES_FILE}>"
)

perform_unit_tests(cpp_library, config)
perform_example_compilation_tests(cpp_library, config)

terminate(true)
